File: sdks/dart/lib/src/collection.dart
/// MongoDB collection for mongo.do
library;

import 'dart:async';
import 'types.dart';
import 'cursor.dart';
import 'selector.dart';

/// A MongoDB collection
class MongoCollection {
  final RpcTransport _transport;
  final String _dbName;
  final String _collectionName;

  MongoCollection(this._transport, this._dbName, this._collectionName);

  /// Get the collection name
  String get collectionName => _collectionName;

  /// Get the database name
  String get databaseName => _dbName;

  /// Find documents matching a filter
  MongoCursor<Document> find([dynamic filter, FindOptions? options]) {
    final filterMap = _normalizeFilter(filter);
    final optionsMap = options?.toJson() ?? <String, dynamic>{};
    return MongoCursor<Document>(
      _transport,
      _dbName,
      _collectionName,
      filterMap,
      optionsMap,
    );
  }

  /// Find a single document matching a filter
  Future<Document?> findOne([dynamic filter, FindOptions? options]) async {
    final filterMap = _normalizeFilter(filter);
    final optionsMap = {...(options?.toJson() ?? {}), 'limit': 1};
    final cursor = MongoCursor<Document>(
      _transport,
      _dbName,
      _collectionName,
      filterMap,
      optionsMap,
    );
    return cursor.first();
  }

  /// Find a document by ID
  Future<Document?> findById(dynamic id) async {
    return findOne(where.id(id));
  }

  /// Insert a single document
  Future<InsertOneResult> insertOne(Document document) async {
    final result = await _transport.call('insertOne', [
      _dbName,
      _collectionName,
      document,
    ]);
    return InsertOneResult.fromJson(result as Map<String, dynamic>);
  }

  /// Insert multiple documents
  Future<InsertManyResult> insertMany(List<Document> documents) async {
    final result = await _transport.call('insertMany', [
      _dbName,
      _collectionName,
      documents,
    ]);
    return InsertManyResult.fromJson(result as Map<String, dynamic>);
  }

  /// Update a single document
  Future<UpdateResult> updateOne(
    dynamic filter,
    dynamic update, [
    UpdateOptions? options,
  ]) async {
    final filterMap = _normalizeFilter(filter);
    final updateMap = _normalizeUpdate(update);
    final result = await _transport.call('updateOne', [
      _dbName,
      _collectionName,
      filterMap,
      updateMap,
      options?.toJson() ?? {},
    ]);
    return UpdateResult.fromJson(result as Map<String, dynamic>);
  }

  /// Update multiple documents
  Future<UpdateResult> updateMany(
    dynamic filter,
    dynamic update, [
    UpdateOptions? options,
  ]) async {
    final filterMap = _normalizeFilter(filter);
    final updateMap = _normalizeUpdate(update);
    final result = await _transport.call('updateMany', [
      _dbName,
      _collectionName,
      filterMap,
      updateMap,
      options?.toJson() ?? {},
    ]);
    return UpdateResult.fromJson(result as Map<String, dynamic>);
  }

  /// Replace a single document
  Future<UpdateResult> replaceOne(
    dynamic filter,
    Document replacement, [
    UpdateOptions? options,
  ]) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('replaceOne', [
      _dbName,
      _collectionName,
      filterMap,
      replacement,
      options?.toJson() ?? {},
    ]);
    return UpdateResult.fromJson(result as Map<String, dynamic>);
  }

  /// Delete a single document
  Future<DeleteResult> deleteOne(dynamic filter, [DeleteOptions? options]) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('deleteOne', [
      _dbName,
      _collectionName,
      filterMap,
      options?.toJson() ?? {},
    ]);
    return DeleteResult.fromJson(result as Map<String, dynamic>);
  }

  /// Delete multiple documents
  Future<DeleteResult> deleteMany(dynamic filter, [DeleteOptions? options]) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('deleteMany', [
      _dbName,
      _collectionName,
      filterMap,
      options?.toJson() ?? {},
    ]);
    return DeleteResult.fromJson(result as Map<String, dynamic>);
  }

  /// Find and update a document atomically
  Future<Document?> findOneAndUpdate(
    dynamic filter,
    dynamic update, {
    bool? upsert,
    bool returnDocument = false,
    Map<String, dynamic>? sort,
    Map<String, dynamic>? projection,
  }) async {
    final filterMap = _normalizeFilter(filter);
    final updateMap = _normalizeUpdate(update);
    final result = await _transport.call('findOneAndUpdate', [
      _dbName,
      _collectionName,
      filterMap,
      updateMap,
      {
        'upsert': upsert,
        'returnDocument': returnDocument ? 'after' : 'before',
        if (sort != null) 'sort': sort,
        if (projection != null) 'projection': projection,
      },
    ]);
    return result as Document?;
  }

  /// Find and delete a document atomically
  Future<Document?> findOneAndDelete(
    dynamic filter, {
    Map<String, dynamic>? sort,
    Map<String, dynamic>? projection,
  }) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('findOneAndDelete', [
      _dbName,
      _collectionName,
      filterMap,
      {
        if (sort != null) 'sort': sort,
        if (projection != null) 'projection': projection,
      },
    ]);
    return result as Document?;
  }

  /// Find and replace a document atomically
  Future<Document?> findOneAndReplace(
    dynamic filter,
    Document replacement, {
    bool? upsert,
    bool returnDocument = false,
    Map<String, dynamic>? sort,
    Map<String, dynamic>? projection,
  }) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('findOneAndReplace', [
      _dbName,
      _collectionName,
      filterMap,
      replacement,
      {
        'upsert': upsert,
        'returnDocument': returnDocument ? 'after' : 'before',
        if (sort != null) 'sort': sort,
        if (projection != null) 'projection': projection,
      },
    ]);
    return result as Document?;
  }

  /// Count documents matching a filter
  Future<int> countDocuments([dynamic filter, CountOptions? options]) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('countDocuments', [
      _dbName,
      _collectionName,
      filterMap,
      options?.toJson() ?? {},
    ]);
    return result as int;
  }

  /// Get an estimated document count
  Future<int> estimatedDocumentCount() async {
    final result = await _transport.call('estimatedDocumentCount', [
      _dbName,
      _collectionName,
    ]);
    return result as int;
  }

  /// Get distinct values for a field
  Future<List<dynamic>> distinct(String field, [dynamic filter]) async {
    final filterMap = _normalizeFilter(filter);
    final result = await _transport.call('distinct', [
      _dbName,
      _collectionName,
      field,
      filterMap,
    ]);
    return (result as List<dynamic>);
  }

  /// Run an aggregation pipeline
  Future<List<Document>> aggregate(
    List<Map<String, dynamic>> pipeline, [
    AggregateOptions? options,
  ]) async {
    final result = await _transport.call('aggregate', [
      _dbName,
      _collectionName,
      pipeline,
      options?.toJson() ?? {},
    ]);
    return (result as List<dynamic>).cast<Document>();
  }

  /// Create an index
  Future<String> createIndex(
    Map<String, dynamic> keys, {
    String? name,
    bool? unique,
    bool? sparse,
    int? expireAfterSeconds,
  }) async {
    final result = await _transport.call('createIndex', [
      _dbName,
      _collectionName,
      keys,
      {
        if (name != null) 'name': name,
        if (unique != null) 'unique': unique,
        if (sparse != null) 'sparse': sparse,
        if (expireAfterSeconds != null) 'expireAfterSeconds': expireAfterSeconds,
      },
    ]);
    return result as String;
  }

  /// Create multiple indexes
  Future<List<String>> createIndexes(List<Map<String, dynamic>> indexes) async {
    final result = await _transport.call('createIndexes', [
      _dbName,
      _collectionName,
      indexes,
    ]);
    return (result as List<dynamic>).cast<String>();
  }

  /// Drop an index
  Future<void> dropIndex(String indexName) async {
    await _transport.call('dropIndex', [
      _dbName,
      _collectionName,
      indexName,
    ]);
  }

  /// Drop all indexes
  Future<void> dropIndexes() async {
    await _transport.call('dropIndexes', [
      _dbName,
      _collectionName,
    ]);
  }

  /// List indexes
  Future<List<Document>> listIndexes() async {
    final result = await _transport.call('listIndexes', [
      _dbName,
      _collectionName,
    ]);
    return (result as List<dynamic>).cast<Document>();
  }

  /// Drop this collection
  Future<void> drop() async {
    await _transport.call('dropCollection', [
      _dbName,
      _collectionName,
    ]);
  }

  /// Rename this collection
  Future<void> rename(String newName, {bool dropTarget = false}) async {
    await _transport.call('renameCollection', [
      _dbName,
      _collectionName,
      newName,
      {'dropTarget': dropTarget},
    ]);
  }

  /// Bulk write operations
  Future<BulkWriteResult> bulkWrite(
    List<Map<String, dynamic>> operations, {
    bool ordered = true,
  }) async {
    final result = await _transport.call('bulkWrite', [
      _dbName,
      _collectionName,
      operations,
      {'ordered': ordered},
    ]);
    return BulkWriteResult.fromJson(result as Map<String, dynamic>);
  }

  /// Watch for changes (change stream)
  Stream<ChangeEvent> watch([List<Map<String, dynamic>>? pipeline]) {
    final controller = StreamController<ChangeEvent>();

    // Implement change stream via polling or WebSocket
    // This is a simplified implementation
    controller.onListen = () async {
      try {
        while (!controller.isClosed) {
          final result = await _transport.call('watch', [
            _dbName,
            _collectionName,
            pipeline ?? [],
          ]);
          if (result != null) {
            controller.add(ChangeEvent.fromJson(result as Map<String, dynamic>));
          }
          await Future.delayed(const Duration(milliseconds: 100));
        }
      } catch (e) {
        if (!controller.isClosed) {
          controller.addError(e);
        }
      }
    };

    return controller.stream;
  }

  Map<String, dynamic> _normalizeFilter(dynamic filter) {
    if (filter == null) return {};
    if (filter is SelectorBuilder) return filter.map;
    if (filter is Map<String, dynamic>) return filter;
    throw ArgumentError('Invalid filter type: ${filter.runtimeType}');
  }

  Map<String, dynamic> _normalizeUpdate(dynamic update) {
    if (update is ModifierBuilder) return update.map;
    if (update is Map<String, dynamic>) return update;
    throw ArgumentError('Invalid update type: ${update.runtimeType}');
  }
}

/// A typed MongoDB collection
class TypedCollection<T> {
  final MongoCollection _collection;
  final T Function(Map<String, dynamic>) _fromJson;
  final Map<String, dynamic> Function(T) _toJson;

  TypedCollection(
    this._collection,
    this._fromJson,
    this._toJson,
  );

  /// Find documents matching a filter
  Future<List<T>> find([dynamic filter, FindOptions? options]) async {
    final cursor = _collection.find(filter, options);
    final docs = await cursor.toList();
    return docs.map(_fromJson).toList();
  }

  /// Find a single document matching a filter
  Future<T?> findOne([dynamic filter, FindOptions? options]) async {
    final doc = await _collection.findOne(filter, options);
    return doc != null ? _fromJson(doc) : null;
  }

  /// Insert a single document
  Future<InsertOneResult> insertOne(T document) async {
    return _collection.insertOne(_toJson(document));
  }

  /// Insert multiple documents
  Future<InsertManyResult> insertMany(List<T> documents) async {
    return _collection.insertMany(documents.map(_toJson).toList());
  }

  /// Update a single document
  Future<UpdateResult> updateOne(
    dynamic filter,
    dynamic update, [
    UpdateOptions? options,
  ]) async {
    return _collection.updateOne(filter, update, options);
  }

  /// Update multiple documents
  Future<UpdateResult> updateMany(
    dynamic filter,
    dynamic update, [
    UpdateOptions? options,
  ]) async {
    return _collection.updateMany(filter, update, options);
  }

  /// Delete a single document
  Future<DeleteResult> deleteOne(dynamic filter, [DeleteOptions? options]) async {
    return _collection.deleteOne(filter, options);
  }

  /// Delete multiple documents
  Future<DeleteResult> deleteMany(dynamic filter, [DeleteOptions? options]) async {
    return _collection.deleteMany(filter, options);
  }

  /// Count documents matching a filter
  Future<int> countDocuments([dynamic filter, CountOptions? options]) async {
    return _collection.countDocuments(filter, options);
  }
}

/// Bulk write result
class BulkWriteResult {
  final int insertedCount;
  final int matchedCount;
  final int modifiedCount;
  final int deletedCount;
  final int upsertedCount;
  final Map<int, dynamic> upsertedIds;

  const BulkWriteResult({
    required this.insertedCount,
    required this.matchedCount,
    required this.modifiedCount,
    required this.deletedCount,
    required this.upsertedCount,
    required this.upsertedIds,
  });

  factory BulkWriteResult.fromJson(Map<String, dynamic> json) {
    final upsertedIdsRaw = json['upsertedIds'] as Map<String, dynamic>?;
    return BulkWriteResult(
      insertedCount: json['insertedCount'] as int? ?? 0,
      matchedCount: json['matchedCount'] as int? ?? 0,
      modifiedCount: json['modifiedCount'] as int? ?? 0,
      deletedCount: json['deletedCount'] as int? ?? 0,
      upsertedCount: json['upsertedCount'] as int? ?? 0,
      upsertedIds: upsertedIdsRaw?.map((k, v) => MapEntry(int.parse(k), v)) ?? {},
    );
  }
}

/// Change event from a change stream
class ChangeEvent {
  final String operationType;
  final Document? fullDocument;
  final Document? documentKey;
  final Document? updateDescription;
  final String? clusterTime;

  const ChangeEvent({
    required this.operationType,
    this.fullDocument,
    this.documentKey,
    this.updateDescription,
    this.clusterTime,
  });

  factory ChangeEvent.fromJson(Map<String, dynamic> json) {
    return ChangeEvent(
      operationType: json['operationType'] as String,
      fullDocument: json['fullDocument'] as Document?,
      documentKey: json['documentKey'] as Document?,
      updateDescription: json['updateDescription'] as Document?,
      clusterTime: json['clusterTime'] as String?,
    );
  }
}

File: sdks/typescript/src/collection.ts
/**
 * Collection class - MongoDB-compatible collection operations
 */

import type {
  Document,
  WithId,
  Filter,
  UpdateFilter,
  FindOptions,
  InsertOneResult,
  InsertManyResult,
  UpdateResult,
  DeleteResult,
  CountDocumentsOptions,
  UpdateOptions,
  ReplaceOptions,
  DeleteOptions,
  AggregationStage,
  AggregateOptions,
  RpcTransport,
} from './types.js';

import { FindCursor, AggregationCursor } from './cursor.js';

/**
 * Collection class providing MongoDB-compatible operations
 */
export class Collection<T extends Document = Document> {
  private _transport: RpcTransport;
  private _dbName: string;
  private _name: string;

  constructor(transport: RpcTransport, dbName: string, name: string) {
    this._transport = transport;
    this._dbName = dbName;
    this._name = name;
  }

  /**
   * Get the collection name
   */
  get collectionName(): string {
    return this._name;
  }

  /**
   * Get the database name
   */
  get dbName(): string {
    return this._dbName;
  }

  /**
   * Get the namespace
   */
  get namespace(): string {
    return `${this._dbName}.${this._name}`;
  }

  // ============================================================================
  // Insert Operations
  // ============================================================================

  /**
   * Insert a single document
   */
  async insertOne(doc: T): Promise<InsertOneResult> {
    const result = await this._transport.call('insertOne', this._dbName, this._name, doc);
    return result as InsertOneResult;
  }

  /**
   * Insert multiple documents
   */
  async insertMany(docs: T[]): Promise<InsertManyResult> {
    const result = await this._transport.call('insertMany', this._dbName, this._name, docs);
    return result as InsertManyResult;
  }

  // ============================================================================
  // Find Operations
  // ============================================================================

  /**
   * Find documents matching a filter - returns a cursor
   */
  find(filter: Filter<T> = {}, options?: FindOptions<T>): FindCursor<WithId<T>> {
    const cursor = new FindCursor<WithId<T>>(this._transport, this._dbName, this._name, filter as Document);

    if (options?.sort) cursor.sort(options.sort as { [key: string]: 1 | -1 });
    if (options?.limit !== undefined) cursor.limit(options.limit);
    if (options?.skip !== undefined) cursor.skip(options.skip);
    if (options?.projection) cursor.project(options.projection);
    if (options?.batchSize) cursor.batchSize(options.batchSize);
    if (options?.maxTimeMS) cursor.maxTimeMS(options.maxTimeMS);
    if (options?.hint) cursor.hint(options.hint);
    if (options?.comment) cursor.comment(options.comment);

    return cursor;
  }

  /**
   * Find a single document matching a filter
   */
  async findOne(filter: Filter<T> = {}, options?: FindOptions<T>): Promise<WithId<T> | null> {
    const cursor = this.find(filter, { ...options, limit: 1 });
    return cursor.next();
  }

  /**
   * Find a document and update it
   */
  async findOneAndUpdate(
    filter: Filter<T>,
    update: UpdateFilter<T>,
    options?: { returnDocument?: 'before' | 'after'; upsert?: boolean }
  ): Promise<WithId<T> | null> {
    const result = await this._transport.call(
      'findOneAndUpdate',
      this._dbName,
      this._name,
      filter,
      update,
      options ?? {}
    );
    return result as WithId<T> | null;
  }

  /**
   * Find a document and delete it
   */
  async findOneAndDelete(filter: Filter<T>): Promise<WithId<T> | null> {
    const result = await this._transport.call('findOneAndDelete', this._dbName, this._name, filter);
    return result as WithId<T> | null;
  }

  /**
   * Find a document and replace it
   */
  async findOneAndReplace(
    filter: Filter<T>,
    replacement: T,
    options?: { returnDocument?: 'before' | 'after'; upsert?: boolean }
  ): Promise<WithId<T> | null> {
    const result = await this._transport.call(
      'findOneAndReplace',
      this._dbName,
      this._name,
      filter,
      replacement,
      options ?? {}
    );
    return result as WithId<T> | null;
  }

  // ============================================================================
  // Update Operations
  // ============================================================================

  /**
   * Update a single document
   */
  async updateOne(filter: Filter<T>, update: UpdateFilter<T>, options?: UpdateOptions): Promise<UpdateResult> {
    const result = await this._transport.call('updateOne', this._dbName, this._name, filter, update, options ?? {});
    return result as UpdateResult;
  }

  /**
   * Update multiple documents
   */
  async updateMany(filter: Filter<T>, update: UpdateFilter<T>, options?: UpdateOptions): Promise<UpdateResult> {
    const result = await this._transport.call('updateMany', this._dbName, this._name, filter, update, options ?? {});
    return result as UpdateResult;
  }

  /**
   * Replace a single document
   */
  async replaceOne(filter: Filter<T>, replacement: T, options?: ReplaceOptions): Promise<UpdateResult> {
    const result = await this._transport.call('replaceOne', this._dbName, this._name, filter, replacement, options ?? {});
    return result as UpdateResult;
  }

  // ============================================================================
  // Delete Operations
  // ============================================================================

  /**
   * Delete a single document
   */
  async deleteOne(filter: Filter<T>, options?: DeleteOptions): Promise<DeleteResult> {
    const result = await this._transport.call('deleteOne', this._dbName, this._name, filter, options ?? {});
    return result as DeleteResult;
  }

  /**
   * Delete multiple documents
   */
  async deleteMany(filter: Filter<T>, options?: DeleteOptions): Promise<DeleteResult> {
    const result = await this._transport.call('deleteMany', this._dbName, this._name, filter, options ?? {});
    return result as DeleteResult;
  }

  // ============================================================================
  // Count Operations
  // ============================================================================

  /**
   * Count documents matching a filter
   */
  async countDocuments(filter: Filter<T> = {}, options?: CountDocumentsOptions): Promise<number> {
    const result = await this._transport.call('countDocuments', this._dbName, this._name, filter, options ?? {});
    return result as number;
  }

  /**
   * Get an estimated document count
   */
  async estimatedDocumentCount(): Promise<number> {
    const result = await this._transport.call('estimatedDocumentCount', this._dbName, this._name);
    return result as number;
  }

  // ============================================================================
  // Aggregation Operations
  // ============================================================================

  /**
   * Run an aggregation pipeline
   */
  aggregate<R extends Document = Document>(
    pipeline: AggregationStage[] = [],
    options?: AggregateOptions
  ): AggregationCursor<R> {
    return new AggregationCursor<R>(
      this._transport,
      this._dbName,
      this._name,
      pipeline as Document[],
      options ?? {}
    );
  }

  /**
   * Get distinct values for a field
   */
  async distinct<K extends keyof T>(field: K, filter?: Filter<T>): Promise<T[K][]> {
    const result = await this._transport.call('distinct', this._dbName, this._name, String(field), filter ?? {});
    return result as T[K][];
  }

  // ============================================================================
  // Index Operations
  // ============================================================================

  /**
   * Create an index
   */
  async createIndex(keys: Document, options?: Document): Promise<string> {
    const result = await this._transport.call('createIndex', this._dbName, this._name, keys, options ?? {});
    return result as string;
  }

  /**
   * Create multiple indexes
   */
  async createIndexes(indexes: Array<{ key: Document; options?: Document }>): Promise<string[]> {
    const result = await this._transport.call('createIndexes', this._dbName, this._name, indexes);
    return result as string[];
  }

  /**
   * Drop an index
   */
  async dropIndex(indexName: string): Promise<void> {
    await this._transport.call('dropIndex', this._dbName, this._name, indexName);
  }

  /**
   * Drop all indexes
   */
  async dropIndexes(): Promise<void> {
    await this._transport.call('dropIndexes', this._dbName, this._name);
  }

  /**
   * List all indexes
   */
  async listIndexes(): Promise<Document[]> {
    const result = await this._transport.call('listIndexes', this._dbName, this._name);
    return result as Document[];
  }

  // ============================================================================
  // Collection Operations
  // ============================================================================

  /**
   * Drop the collection
   */
  async drop(): Promise<boolean> {
    const result = await this._transport.call('dropCollection', this._dbName, this._name);
    return result as boolean;
  }

  /**
   * Rename the collection
   */
  async rename(newName: string, options?: { dropTarget?: boolean }): Promise<void> {
    await this._transport.call('renameCollection', this._dbName, this._name, newName, options ?? {});
    this._name = newName;
  }

  // ============================================================================
  // Bulk Operations
  // ============================================================================

  /**
   * Perform bulk write operations
   */
  async bulkWrite(
    operations: Array<
      | { insertOne: { document: T } }
      | { updateOne: { filter: Filter<T>; update: UpdateFilter<T>; upsert?: boolean } }
      | { updateMany: { filter: Filter<T>; update: UpdateFilter<T>; upsert?: boolean } }
      | { deleteOne: { filter: Filter<T> } }
      | { deleteMany: { filter: Filter<T> } }
      | { replaceOne: { filter: Filter<T>; replacement: T; upsert?: boolean } }
    >,
    options?: { ordered?: boolean }
  ): Promise<{
    insertedCount: number;
    matchedCount: number;
    modifiedCount: number;
    deletedCount: number;
    upsertedCount: number;
    upsertedIds: Record<number, string>;
  }> {
    const result = await this._transport.call('bulkWrite', this._dbName, this._name, operations, options ?? {});
    return result as {
      insertedCount: number;
      matchedCount: number;
      modifiedCount: number;
      deletedCount: number;
      upsertedCount: number;
      upsertedIds: Record<number, string>;
    };
  }
}
